File: week_2/day_08_stack/valid_parentheses_stack.py
# Program: valid_parentheses_stack.py

def isValid(s):
    stack = []
    mapping = {')': '(', ']': '[', '}': '{'}

    for ch in s:
        if ch in mapping:
            if not stack or stack.pop() != mapping[ch]:
                return False
        else:
            stack.append(ch)

    return not stack


# INPUT
s = "()[]{}"
print(isValid(s))   # Output: True

s2 = "(]"
print(isValid(s2))  # Output: False

File: week_2/day_08_stack/validParenthesesStack.js
// Program: validParenthesesStack.js

function isValid(s) {
    const stack = [];
    const map = { ')': '(', ']': '[', '}': '{' };

    for (let ch of s) {
        if (map[ch]) {
            if (!stack.length || stack.pop() !== map[ch]) {
                return false;
            }
        } else {
            stack.push(ch);
        }
    }
    return stack.length === 0;
}

// INPUT
console.log(isValid("()[]{}")); // true
console.log(isValid("(]"));     // false
